cmd/relay/relay.go

```diff
@@ -169,3 +155,132 @@ func Run(ctx context.Context, config Config) error {
 		}
 	}
 }
+
+// newENRHandler returns a handler that returns the nodes ID and public address encoded as a ENR.
+func newENRHandler(ctx context.Context, tcpNode host.Host, p2pKey *ecdsa.PrivateKey, config p2p.Config) func(ctx context.Context) ([]byte, error) {
+	// Resolve external hostname peridodically.
+	var (
+		extHostMu sync.Mutex
+		extHostIP net.IP
+	)
+	go func() {
+		if config.ExternalHost == "" {
+			return
+		}
+
+		resolveExtHost := func() {
+			ip, err := net.LookupIP(config.ExternalHost)
+			if err != nil {
+				log.Warn(ctx, "Failed to resolve external host", err, z.Str("host", config.ExternalHost))
+				return
+			}
+			extHostMu.Lock()
+			extHostIP = ip[0]
+			extHostMu.Unlock()
+		}
+
+		onStartup := make(chan struct{}, 1)
+		onStartup <- struct{}{}
+
+		for {
+			select {
+			case <-onStartup:
+				resolveExtHost()
+			case <-time.After(5 * time.Minute):
+				resolveExtHost()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// getExtHostIP returns the external host IP and true if it is set.
+	getExtHostIP := func() (net.IP, bool) {
+		extHostMu.Lock()
+		defer extHostMu.Unlock()
+
+		return extHostIP, len(extHostIP) != 0
+	}
+
+	return func(ctx context.Context) ([]byte, error) {
+		// Use libp2p configured and detected addresses.
+		addrs := tcpNode.Addrs()
+		if len(addrs) == 0 {
+			return nil, errors.New("no addresses")
+		}
+
+		// Order public addresses first.
+		sort.SliceStable(addrs, func(i, j int) bool {
+			iPublic, jPublic := manet.IsPublicAddr(addrs[i]), manet.IsPublicAddr(addrs[j])
+			if jPublic && !iPublic {
+				return true // Only swap if j is public and i is not.
+			}
+
+			return false
+		})
+
+		// Use first address (ip and port).
+		addr, err := manet.ToNetAddr(addrs[0])
```

Comment:
discuss: why are we using only the first address?

Reply:
ENRs only supports 1 address